// @claude-flow/embeddings/dist/index.js
/**
 * V3 Embedding Service Module
 *
 * Production embedding service aligned with agentic-flow@alpha:
 * - OpenAI provider (text-embedding-3-small/large)
 * - Transformers.js provider (local ONNX models)
 * - Agentic-flow provider (optimized ONNX with SIMD)
 * - Mock provider (development/testing)
 *
 * Additional features:
 * - Persistent SQLite cache
 * - Document chunking with overlap
 * - L2/L1/minmax/zscore normalization
 * - Hyperbolic embeddings (Poincaré ball)
 * - Neural substrate integration (drift, memory, swarm)
 *
 * @module @claude-flow/embeddings
 */
export * from './types.js';
export * from './embedding-service.js';
// Re-export commonly used items at top level
export { createEmbeddingService, createEmbeddingServiceAsync, getEmbedding, cosineSimilarity, euclideanDistance, dotProduct, computeSimilarity, OpenAIEmbeddingService, TransformersEmbeddingService, MockEmbeddingService, AgenticFlowEmbeddingService, } from './embedding-service.js';
// Chunking utilities
export { chunkText, estimateTokens, reconstructFromChunks, } from './chunking.js';
// Normalization utilities
export { l2Normalize, l2NormalizeInPlace, l1Normalize, minMaxNormalize, zScoreNormalize, normalize, normalizeBatch, l2Norm, isNormalized, centerEmbeddings, } from './normalization.js';
// Hyperbolic embeddings (Poincaré ball)
export { euclideanToPoincare, poincareToEuclidean, hyperbolicDistance, mobiusAdd, mobiusScalarMul, hyperbolicCentroid, batchEuclideanToPoincare, pairwiseHyperbolicDistances, isInPoincareBall, } from './hyperbolic.js';
// Persistent cache
export { PersistentEmbeddingCache, isPersistentCacheAvailable, } from './persistent-cache.js';
// Neural substrate integration
export { NeuralEmbeddingService, createNeuralService, isNeuralAvailable, listEmbeddingModels, downloadEmbeddingModel, } from './neural-integration.js';
